Added Template Runner

matthew-rollitt opened the pull request:
Small program that can consume a json file and create and tests jobs/pools on them. If the job fails the container should be left over

matthew-rollitt committed:
Added Template Runner

Small program that can consume a json file and create and tests jobs/pools on them. If the job fails the container should be left over

Tatsinnit commented:
💡 📔   @matthew-rollitt : Please write context of this work in detail for the Reviewers to understand the file reference and their importance. And little gist for this work as well please.

* Before merging you can summaries it.

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,154 @@
+import azure.storage.blob as azureblob
```
:bulb: **Suggestion: Minor please take care moving forward with final version** Unit Tests? for this entire file will be great idea. All the public functs should hooked up to UTs

matthew-rollitt replied:
But isn't this a testing program, If I do that wouldn't I be writing UTs for UTs?

Tatsinnit replied:
🐛 This seems not true, You have **No** `asserts` inside `JobRunner` ? if this is true then UTs will usually have some sort of `assert` of some sort. 

* Also if the intention is to make `jobrunner` a UT then you should follow a convetion of starting file as: `Test_` 

But please take that as a note: Reason behind UT is as follows:
* This will introduce better file convention in your project.
* Also JobRunner has no `asserts` so definetly this is some sort of runner program, having test makes it easy for you in longer run. :)

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,154 @@
+import time
+import sys
+import traceback
+import asyncio
```
:bulb: **minor** Same import multiple times: see line : 17

matthew-rollitt replied:
removed unneeded imports

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,154 @@
+import random
+import types
+import functools  
+
```
🗒 **minor** note:  you can decide what length will be good for this work, but a specific standard will help in longer run for this project. some of PEP 8 here: https://stackoverflow.com/questions/88942/why-does-pep-8-specify-a-maximum-line-length-of-79-characters

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,278 @@
+    return batchmodels.ResourceFile(file_path=blob_name,
+                                    blob_source=sas_url)
+
+async def wait_for_tasks_to_complete(batch_service_client, job_id, timeout):
```
:bulb: 🐛 : what is the `async` behavior this funct is trying to do? **apart** from `async.sleep` on line242` which seems like might be related to any scenario? also this returns the tuple or some sort. i.e. `True or False, success or value error.`

Tatsinnit replied:
For example: Please see your `RunJob` implementation which has await on `submit_job` which does has await on `batch_service_client.job.add, job`

matthew-rollitt replied:
We don't need to make this async since we are waiting for the task to complete.

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,278 @@
+    while datetime.datetime.now() < timeout_expiration:
+        tasks = batch_service_client.task.list(job_id)
+
+        incomplete_tasks = [task for task in tasks if
```
📓 💡 Please note: this will return list of `incomplete_tasks` and then eventually **no operation** is happening on this list? and code returns with True or false, is this correct?

matthew-rollitt replied:
Correct, we are using this code to check to see if all the tasks are in a completed state. We check the output of the task in 'check_task_output'

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,278 @@
+    return False, ValueError("ERROR: Tasks did not reach 'Completed' state within "
+                       "timeout period of " + str(timeout))
+
+async def check_task_output(batch_service_client, job_id, expected_output):
```
💡 **Important** Please note: this is `async` method but actual operation within this method is all non async, i.e. what is awaitable in this method? [helpful link](https://docs.python.org/3/library/asyncio.html) if this funct will get yielded somehwere above this funct needs to have async behavior. _Please refer to a sample I created_.

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,278 @@
+    try:
+        template["parameters"]["jobId"]["defaultValue"] = job_id
+    except KeyError:
+        pass        
```
💡 **Minor** Is this intentional?

matthew-rollitt replied:
yep, from what I was seeing online the best way to handle it when a json value doesn't exist is just except the error and move on

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,154 @@
+    job_json = batch_service_client.job.expand_template(template, parameters)
+    job = batch_service_client.job.jobparameter_from_json(job_json)
+    loop = asyncio.get_event_loop()
+    await loop.run_in_executor(None, functools.partial(batch_service_client.job.add, job))    
```
🏗 **Something to think** What is the `returntype` of this?

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,278 @@
+    """
+    
+    #print('Printing task output...')
+
```
💡 **minor** This might help further: https://www.python.org/dev/peps/pep-0492/  also if you wish to differentiate with `private` method this link might come handy: https://stackoverflow.com/questions/17193457/private-methods-in-python

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,154 @@
+
+        
+
+    async def validate(self, batch_service_client):
```
💡 **Important**

**Sample for better understanding** this is just vanilla i created and you can play around with the idea. :)

![image](https://user-images.githubusercontent.com/6233295/46061226-167a6d00-c1ba-11e8-84fa-c026910aee97.png)



```
import asyncio
async def get_programmer_name(name):
    await asyncio.sleep(3)
    return "programmer-%s" % name
async def foobar():
    result = await get_programmer_name("matt")
    print("---return {}r---".format(result))
    return result
def foomain():
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    result = loop.run_until_complete(foobar())
    return result
programmer_name = foomain()
---return programmer-mattr---
```

Tatsinnit replied:
:bulb: **Important** Please note: Sample below is from async to Blocking call: using `run_in_executor`:

![image](https://user-images.githubusercontent.com/6233295/46119318-804c5280-c25e-11e8-83f2-e971fa7fabff.png)

```
def blocking_io(arg1="hulk", arg2="smash"):
    monty = "{}{}".format(arg1, arg2)
    print("----- {}".format(monty))
    return monty
async def async_foo():
    loop = asyncio.get_event_loop()
    result = await loop.run_in_executor(
        None, blocking_io, "IronMan")
    print('default thread pool', result)
    return result
def foo():
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    result = loop.run_until_complete(async_foo())
    print("returned result is - {}".format(result))
    return result
avengers = foo()
----- IronMansmash
default thread pool IronMansmash
returned result is - IronMansmash

```

Tatsinnit replied:
📔  -> follow up ping on this, is this sorted, I checked recent commit seems like not catering the exact use of async used between : `main call -> to Validate -> Wait` method. Play around and please let me know and we can sit down and make small changes.

💡 : Future reference only: Best for these scenarios moving forwards will be make small test files add UTs for your scenario so that your functs are contolled.

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,154 @@
+    def __str__(self):
+        return "job_id: {} pool_id:{} ".format(self.job_id, self.pool_id)
+
+    async def RunJob(self, batch_service_client):
```
💡 📓 **Minor recommendation**: Python now support `type annotations` so that can help in specifying type of input to the function and its return types. For longer run this will help in organical growth of this file, and Unit test writeups become easier: for example:

```
async def monty(
   self,
   obj_1: str
   obj_2: some_type
) -> return_type :
   funct_body
   pass
```

matthew-rollitt replied:
Thanks I'll keep that in mind

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,120 @@
+import azure.batch.models.batch_error 
+import azure.common
+import logging
+import websockets
```
💡 📓 **Future good practice** Some of these are unused imports, I would encourage to use `flake 8` its very imple to run on your file and it will catch few things for you. Flake 8 link : http://flake8.pycqa.org/en/latest/

matthew-rollitt replied:
Thanks, I'll use flake 8. I was wondering what the best tool is for the job.

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,120 @@
+            test = template["tests"][i]
+            applicationLicenses = None
+
+            try:
```
💡 **Important** do we need `try - exept` here? seems like this is assignment operation?

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,120 @@
+from __future__ import print_function
```
💡 **Recommendation**: Unt tests? they will come very handy in longer run.

matthew-rollitt replied:
I was thinking about adding tests to this and yes I do like adding tests to this. But in this case, I feel the purpose of this script is really just to test out all the templates. I don't feel like it's worth adding

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,278 @@
+            return True, "job: {} successfully completed.".format(job_id)
+        else:
+            print("job: {} is running".format(job_id))
+            await asyncio.sleep(1)
```
💡**Important** Is this present only to make this fucntion `async`? or any contextual reason for `sleep(1)` here for this method?

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,278 @@
+    for task in tasks:    
+        all_files = batch_service_client.file.list_from_task(job_id, task.id, recursive=True)
+        for f in all_files:
+            if expected_output in f.name:
```
💡 **Important** In this context is `f.name` a list or just `string with finlename` if latter is the case then you can just do `==` instead of `in` use.

matthew-rollitt replied:
need to use in since we to deal with the filepath in the file

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,120 @@
+            try:
+                applicationLicenses = test["applicationLicense"]
+            except KeyError:
+                pass
```
💡 **Minor: just idea share**: you can either return some sort of `invalid` status here or some exit mechanism, using `pass` will jsut silently run the whole on on except `keyError` in this instance. There are few other places with similar pattern.

matthew-rollitt replied:
I can just drop the KeyError in the try/catch. I'm wanting to just continue on because that json element can be empty by design.

matthew-rollitt commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,154 @@
+        
+        # load the template and parameters file 
+        template = ""
+        with open(self.template_file) as f: 
```
yep can do

matthew-rollitt committed:
Updated async and a few other things

Fixed the async behavior that was causing null errors. Fixed json files so containers can be created correctly. Removed a bunch of Libs that won't be needed and cleaned up a few other methods based on a PR

Tatsinnit commented:
💡 Reference Note: We will not merge this, mainly intention was the first pass, so once main issue is sorted we can close this PR, :) setting comment here, in case anyone accidently merges.
